src/cargo/util/config/mod.rs

```diff
@@ -1176,18 +1176,51 @@ impl Config {
                 CV::Table(map, Definition::Cli)
             } else {
-                // TODO: This should probably use a more narrow parser, reject
-                // comments, blank lines, [headers], etc.
-                let toml_v: toml::Value = toml::de::from_str(arg)
-                    .with_context(|| format!("failed to parse --config argument `{}`", arg))?;
-                let toml_table = toml_v.as_table().unwrap();
-                if toml_table.len() != 1 {
+                // We only want to allow "dotted key" (see https://toml.io/en/v1.0.0#keys)
+                // expressions followed by a value that's not an "inline table"
+                // (https://toml.io/en/v1.0.0#inline-table). Easiest way to check for that is to
+                // parse the value as a toml_edit::Document, and check that the (single)
+                // inner-most table is set via dotted keys.
+                let doc: toml_edit::Document = arg.parse().with_context(|| {
+                    format!("failed to parse value from --config argument `{}` as a dotted key expression", arg)
+                })?;
+                let ok = {
+                    let mut got_to_value = false;
+                    let mut table = doc.as_table();
+                    let mut is_root = true;
+                    while table.is_dotted() || is_root {
+                        is_root = false;
+                        if table.len() != 1 {
+                            break;
+                        }
+                        let (_, n) = table.iter().next().expect("len() == 1 above");
+                        if let Some(nt) = n.as_table() {
+                            table = nt;
+                        } else if n.is_inline_table() {
+                            bail!(
+                                "--config argument `{}` sets a value to an inline table, which is not accepted",
+                                arg,
+                            );
+                        } else {
+                            got_to_value = true;
+                            break;
+                        }
+                    }
+                    got_to_value
+                };
+                if !ok {
                     bail!(
-                        "--config argument `{}` expected exactly one key=value pair, got {} keys",
-                        arg,
-                        toml_table.len()
+                        "--config argument `{}` was not a TOML dotted key expression (a.b.c = _)",
+                        arg
                     );
                 }
+
+                // Rather than trying to bridge between toml_edit::Document and toml::Value, we
+                // just parse the whole argument again now that we know it has the right format.
+                let toml_v = toml::from_str(arg).with_context(|| {
+                    format!("failed to parse value from --config argument `{}`", arg)
+                })?;
```

Comment:
Now that #10086 is merged, we can use `toml::from_document(&doc)` rather than re-parsing the TOML

Reply:
Ah, good catch! Fixed.